gobblin-modules/google-ingestion/src/main/java/gobblin/ingestion/google/webmaster/GoogleWebmasterDataFetcherImpl.java

```diff
@@ -0,0 +1,366 @@
+      throws IOException {
+
+    int startRow = 0;
+    int count = 4;
+    final ExecutorService es = Executors.newFixedThreadPool(count);
+    int r = 0;
+    while (r < 50) {
+      r++;
+      List<Future<Integer>> results = new ArrayList<>(count);
+
+      for (int i = 0; i < count; ++i) {
+        startRow += GoogleWebmasterClient.API_ROW_LIMIT;
+        final int start = startRow;
+        Future<Integer> submit = es.submit(new Callable<Integer>() {
+          @Override
+          public Integer call() {
+            LOG.info(String.format("Getting page size from %s...", start));
+            while (true) {
+              try {
+                List<String> pages =
+                    _client.getPages(_siteProperty, startDate, endDate, country, GoogleWebmasterClient.API_ROW_LIMIT,
+                        requestedDimensions, apiDimensionFilters, start);
+                if (pages.size() < GoogleWebmasterClient.API_ROW_LIMIT) {
+                  //Figured out the size
+                  return pages.size() + start;
+                } else {
+                  //The size is not determined.
+                  return -1;
+                }
+              } catch (IOException e) {
+                LOG.info(String.format("Getting page size from %s failed. Retrying...", start));
+              }
+              try {
+                Thread.sleep(200);
+              } catch (InterruptedException e) {
+                LOG.error(e.getMessage());
+              }
+            }
+          }
+        });
+        results.add(submit);
+        try {
+          Thread.sleep(250);
+        } catch (InterruptedException e) {
+          LOG.error(e.getMessage());
+        }
+      }
+
+      for (Future<Integer> result : results) {
+        try {
+          Integer integer = result.get(2, TimeUnit.MINUTES);
+          if (integer > 0) {
+            es.shutdownNow();
+            return integer;
+          }
+        } catch (InterruptedException e) {
+          throw new RuntimeException(e);
+        } catch (ExecutionException e) {
+          throw new RuntimeException(e);
+        } catch (TimeoutException e) {
+          throw new RuntimeException(e);
+        }
+      }
+    }
+    throw new RuntimeException("Something seems wrong here. Having more than 5000*4*50 pages?");
+  }
+
+  /**
+   * Get all pages in an async mode.
+   */
+  private Collection<String> getPages(String startDate, String endDate, List<Dimension> dimensions,
+      ApiDimensionFilter countryFilter, Queue<Pair<String, FilterOperator>> toProcess) throws IOException {
+    String country = GoogleWebmasterFilter.countryFilterToString(countryFilter);
+
+    ConcurrentLinkedDeque<String> allPages = new ConcurrentLinkedDeque<>();
+    Random random = new Random();
+    //we need to retry many times because the path may be very long
+    final int retry = 120;
+    int r = 0;
+    while (r <= retry) {
```

Comment:
This seems to be fairly complicated with a few arbitrary constants. I don't also think that the concept of rounds is necessary. I suggest having a look at https://github.com/google/guava/wiki/ListenableFutureExplained .

Here is one option:
- You keep two global counters: runningJobs and failedJobs and a Condition variable jobFinished.
- Convert the implementation submitJob to a Callable or Runnable object with all current parameters plus a retry counter. The object also implements FutureCallback:
     - onSuccess, if there are new jobs to be submitted, it will call submitJob() recursively for them; then it decreases the runningJobs counter; signal jobFinished
     - onFailure, it will increase the retry counter; if too, many retries, it will increase the failedJobs counter, decrease the running jobsCounter. If there are more, retries, it re-submits the job; signal jobFinished
- submitJob itself will:
   - Increase the runningJobs counter
   - Submit a ListenableFuture with the above Callable/Runnable and add the future as a callback to itself.
- getPages() submits the initial job(s) to the executor service and then just waits on the condition jobFinished until either 0 == runningJobs or 0 < failedJobs .